### PR TITLE
Load configuration options from workspace root

### LINKED ISSUE #4249: Combine `uv` configuration from workspace root
Rather than the nearest `pyproject.toml`. We lock the whole workspace at once, so I don't think we can allow per-project configuration within a workspace... Perhaps for sync / install, though?

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

In a workspace, we now read configuration from the workspace root. Previously, we read configuration from the first `pyproject.toml` or `uv.toml` file in path -- but in a workspace, that would often be the _project_ rather than the workspace configuration.

We need to read configuration from the workspace root, rather than its members, because we lock the workspace globally, so all configuration applies to the workspace globally.

As part of this change, the `uv-workspace` crate has been renamed to `uv-settings` and its purpose has been narrowed significantly (it no longer discovers a workspace; instead, it just reads the settings from a directory).

If a user has a `uv.toml` in their directory or in a parent directory but is _not_ in a workspace, we will still respect that use-case as before.

Closes #4249.

### CODE REVIEW

File: crates/uv/src/main.rs [hunk 1]
```diff
@@ -117,17 +118,23 @@ async fn run() -> Result<ExitStatus> {
...
     // Load the workspace settings, prioritizing (in order):
     // 1. The configuration file specified on the command-line.
-    // 2. The configuration file in the current directory.
-    // 3. The user configuration file.
+    // 2. The configuration file in the current workspace (i.e., the `pyproject.toml` or `uv.toml`
+    //    file in the workspace root directory). If found, this file is combined with the user
+    //    configuration file.
+    // 3. The nearest `uv.toml` file in the directory tree, starting from the current directory. If
+    //    found, this file is combined with the user configuration file. In this case, we don't
+    //    search for `pyproject.toml` files, since we're not in a workspace.
```
Comment (charliermarsh) [inline crates/uv/src/main.rs:126]: Rules are explained here, in more detail.

**Comment by charliermarsh** [comment]
It's a bit confusing that configuration in a workspace member will effectively be ignored. Something to revisit in the future, I think.

**Comment by zanieb** [comment]
> It's a bit confusing that configuration in a workspace member will effectively be ignored. Something to revisit in the future, I think.

That seems pretty problematic :D

File: crates/uv/tests/pip_compile.rs [hunk 2]
```diff
@@ -3200,38 +3200,20 @@ fn override_dependency_from_workspace_invalid_syntax() -> Result<()> {
...
-        # via
-        #   jinja2
-        #   werkzeug
-    werkzeug==3.0.1
-        # via flask
 
     ----- stderr -----
-    warning: Failed to parse `pyproject.toml`: TOML parse error at line 9, column 29
+    error: Failed to parse: `pyproject.toml`
+      Caused by: TOML parse error at line 9, column 29
```
Comment (zanieb) [inline crates/uv/tests/pip_compile.rs:3209]: Is this behavior change intentional?

**Review by zanieb — APPROVED** [review]

File: crates/uv/tests/pip_compile.rs (hunk 2 above)
Comment (charliermarsh) [reply crates/uv/tests/pip_compile.rs:3209]: Yes, if you have an error in your `override-dependencies` we should fail. The previous behavior was wrong IMO.

**Comment by charliermarsh** [comment]
I know it's a hard question, but what do you think the semantics should be? None of the options are enforceable on a per-package basis.

File: crates/uv/tests/pip_compile.rs (hunk 2 above)
Comment (zanieb) [reply crates/uv/tests/pip_compile.rs:3209]: Makes sense to me

**Comment by zanieb** [comment]
I'm not sure yet. I think I need to play with a real workspace setup first. I think we'd need to be able to map a tree of dependencies to their source project in the workspace to be able to enforce things, right? It seems like it would require a lot of changes.

**Pull request merged**